File: OpenBid/Mopub/ObjectiveC/SampleApp/SampleApp/EventHandlers/MoPub/Banner/MoPubBannerEventHandler.m


#import "MoPubBannerEventHandler.h"
#import <MPError.h>

@interface MoPubBannerEventHandler()<MPAdViewDelegate>
@property (nonatomic, strong) MPAdView *bannerView;
@property (nonatomic, weak) id<POBBannerEventDelegate> delegate;
@property (nonatomic, assign) CGSize adContentSize;
@property (nonatomic, assign) CGSize adSize;
@end

@implementation MoPubBannerEventHandler

-(instancetype)initWithAdUnitId:(NSString *)adUnitId adSize:(CGSize)size{
    self = [super init];
    if (self) {
        
        // Do any additional setup required for MoPub Banner Ads
        // Create a MoPub Banner
        self.bannerView = [[MPAdView alloc] initWithAdUnitId:adUnitId];
        self.bannerView.maxAdSize = size;
        self.adSize = self.bannerView.adContentViewSize;
        
        // Set banner view frame
        CGRect frame = self.bannerView.frame;
        frame.size = self.adSize;
        self.bannerView.frame = frame;
        
        // Set delegates on MPAdView instance, these should not be removed/overridden else event handler will not work as expected.
        self.bannerView.delegate = self;
        
        // Disable MPAdView refresh, refresh will be managed by OpenBid SDK.
        [self.bannerView stopAutomaticallyRefreshingContents];
    }
    return self;
}

-(void)dealloc{
    _bannerView.delegate = nil;
    _bannerView = nil;
    _delegate = nil;
}

#pragma mark - POBBannerEvent delegates
- (CGSize)adContentSize {
    return _adContentSize;
}

- (void)requestAdWithBid:(POBBid *)bid {
    
    _bannerView.keywords = nil;

    // If bid is valid, add bid related keywords on Mopub view
    if (bid) {
        if (self.configBlock) {
            self.configBlock(self.bannerView);
        }
        NSString * keywords = [self keywordsForBid:bid];
        self.bannerView.keywords = keywords;
        NSLog(@"MoPub banner keywords: %@", keywords);

        if (bid.status.boolValue) {
            NSMutableDictionary *localExtras = [NSMutableDictionary dictionaryWithDictionary:self.bannerView.localExtras];
             localExtras[@"pob_bid"] = bid;
            self.bannerView.localExtras = localExtras;
        }
    }
    // NOTE: Please do not remove this code. Need to reset MoPub banner delegate to MoPubBannerEventHandler as these are used by MoPubBannerEventHandler internally. Changing the mopub delegate to other instance may break the callbacks and the banner refresh mechanism.
    if (self.bannerView.delegate != self) {
        NSLog(@"Resetting MoPub banner delegate to MoPubBannerEventHandler as these are used by MoPubBannerEventHandler internally.");
        self.bannerView.delegate = self;
    }
    [self.bannerView loadAd];
}

- (NSArray<POBAdSize *> *)requestedAdSizes{
    POBAdSize *size = POBAdSizeMakeFromCGSize(self.adSize);
    return @[size];
}

- (void)setDelegate:(id<POBBannerEventDelegate>)delegate {
    _delegate = delegate;
}

#pragma mark - MPAdViewDelegate
- (UIViewController *)viewControllerForPresentingModalView {
    return [self.delegate viewControllerForPresentingModal];
}

- (void)adViewDidLoadAd:(MPAdView *)view adSize:(CGSize)adSize {
    _adContentSize = adSize;
    [self.delegate adServerDidWin:view];
}

- (void)adView:(MPAdView *)view didFailToLoadAdWithError:(NSError *)error {
    NSError *eventError = nil;
    switch (error.code) {
        case MOPUBErrorNoNetworkData:
        case MOPUBErrorNoInventory:
        case MOPUBErrorAdapterHasNoInventory:
            // No data found in NSHTTPURL response
            eventError = [self eventError:error withErrorCode:POBErrorNoAds];
            break;
            
        case MOPUBErrorNetworkTimedOut:
            eventError = [self eventError:error withErrorCode:POBErrorNetworkError];
            break;
            
        case MOPUBErrorServerError:
            eventError = [self eventError:error withErrorCode:POBErrorServerError];
            break;
            
        case MOPUBErrorAdRequestTimedOut:
            eventError = [self eventError:error withErrorCode:POBErrorTimeout];
            break;
            
        case MOPUBErrorAdapterInvalid:
        case MOPUBErrorAdapterNotFound:
            eventError = [self eventError:error withErrorCode:POBSignalingError];
            break;
            
        case MOPUBErrorAdUnitWarmingUp:
        case MOPUBErrorSDKNotInitialized:
            eventError = [self eventError:error withErrorCode:POBErrorInternalError];
            break;
            
        case MOPUBErrorUnableToParseJSONAdResponse:
        case MOPUBErrorUnexpectedNetworkResponse:
            eventError = [self eventError:error withErrorCode:POBErrorInvalidResponse];
            break;
            
        case MOPUBErrorNoRenderer:
            eventError = [self eventError:error withErrorCode:POBErrorRenderError];
            break;
            
        default:
            eventError = error;
            break;
    }
    
    [self.delegate failedWithError:eventError];
}

-(void)willPresentModalViewForAd:(MPAdView *)view{
    [self.delegate willPresentModal];
}

-(void)didDismissModalViewForAd:(MPAdView *)view{
    [self.delegate didDismissModal];
}

-(void)willLeaveApplicationFromAd:(MPAdView *)view{
    [self.delegate willLeaveApp];
}

#pragma mark - Private Methods

-(NSError *)eventError:(NSError *)error  withErrorCode:(POBErrorCode )code{
    NSError *eventError = [NSError errorWithDomain:kPOBErrorDomain
                                              code:code
                                          userInfo:error.userInfo];
    return eventError;
}

-(NSString *)keywordsForBid:(POBBid *)bid{
    NSMutableString *keywords = [NSMutableString new];
    NSDictionary *targetingInfo = bid.targetingInfo;
    for (NSString *key in targetingInfo) {
        [keywords appendFormat:@"%@:%@,", key, targetingInfo[key]];
    }
    if (self.bannerView.keywords.length) {
        [keywords appendFormat:@"%@,", self.bannerView.keywords];
    }
    if (keywords.length) {
        [keywords deleteCharactersInRange:NSMakeRange(keywords.length-1, 1)];
    }
    return [NSString stringWithString:keywords];
}

@end

File: OpenWrap/Basic/Mopub/Swift/SampleApp/SampleApp/EventHandlers/MoPub/Banner/MoPubBannerEventHandler.swift
//This class is compatible with OpenWrap SDK v1.5.0

import UIKit

/*!
 This class is responsible for communication between OpenWrap banner view and Mopub banner view.
 It implements the POBBannerEvent protocol. it notifies event back to OpenWrap SDK using POBBannerEventDelegate methods
 */
class MoPubBannerEventHandler: NSObject,POBBannerEvent,MPAdViewDelegate {
    
    /*!
     A configureBlock that is called before event handler makes ad request call to DFP SDK. It passes DFPBannerView & DFPRequest which will be used to make ad request.
     */
    var configBlock: ((_ view: MPAdView)->Void)?

    var bannerView:MPAdView?
    weak var _delegate: POBBannerEventDelegate?
    var contentSize:CGSize?
    var adSize:CGSize!

    /*!
     Initializes and returns a event handler with the specified Mopub ad unit and ad sizes
     
     @param adUnitId Mopub ad unit id
     @param size ad size
     */
    init(adUnitId:String, size:CGSize) {
        super.init()
        // Create a MoPub Banner
        bannerView = MPAdView(adUnitId: adUnitId)
        bannerView?.maxAdSize = size
        adSize = bannerView?.adContentViewSize()
        
        // Set banner view frame
        var frame = bannerView?.frame
        frame?.size = size
        bannerView?.frame = frame ?? CGRect.zero
        
        // Set delegates on MPAdView instance, these should not be removed/overridden else event handler will not work as expected.
        bannerView?.delegate = self
        
        // Disable MPAdView refresh, refresh will be managed by OpenWrap SDK.
        bannerView?.stopAutomaticallyRefreshingContents()
    }
    
    deinit {
        bannerView?.delegate = nil
        bannerView = nil
        _delegate = nil
    }

// MARK: - POBBannerEvent delegates
    func requestAd(with bid: POBBid?) {
        
        bannerView?.keywords = nil;

        // If bid is valid, add bid related keywords on Mopub view
        if bid != nil {
            if (configBlock != nil) {
                configBlock!(bannerView!)
            }

            let keywords = self.keywords(for: bid)
            bannerView?.keywords = keywords
            
            print("MoPub banner keywords: \(String(describing: keywords))")

            if (bid?.status != 0) {
                var localExtras = bannerView?.localExtras ?? [AnyHashable:Any]()
                localExtras["pob_bid"] = bid
                bannerView?.localExtras = localExtras
            }
        }
        // NOTE: Please do not remove this code. Need to reset MoPub banner delegate to MoPubBannerEventHandler as these are used by MoPubBannerEventHandler internally. Changing the mopub delegate to other instance may break the callbacks and the banner refresh mechanism.
        if !(bannerView?.delegate is MoPubBannerEventHandler) {
            NSLog("Resetting MoPub banner delegate to MoPubBannerEventHandler as these are used by MoPubBannerEventHandler internally.");
            bannerView?.delegate = self
        }
        bannerView?.loadAd()
    }
    
    func setDelegate(_ delegate: POBBannerEventDelegate) {
        _delegate = delegate
    }
    
    func adContentSize() -> CGSize {
        return contentSize ?? CGSize.zero
    }
    
    func requestedAdSizes() -> [POBAdSize] {
        return [POBAdSize(cgSize: adSize)]
    }

// MARK: - MPAdViewDelegate
    func viewControllerForPresentingModalView() -> UIViewController! {
        return _delegate?.viewControllerForPresentingModal()
    }
    
    func adViewDidLoadAd(_ view: MPAdView!, adSize: CGSize) {
        contentSize = adSize
        let bid = self.bannerView?.localExtras["pob_bid"] as? POBBid
        if bid!.hasWon {
            _delegate?.openWrapPartnerDidWin()
        }
        else {
            _delegate?.adServerDidWin(view)
        }
    }
    
    func adView(_ view: MPAdView!, didFailToLoadAdWithError error: Error!) {
        var eventErr = error as NSError
        let mopubError = MOPUBErrorCode(rawValue: Int32(eventErr.code))
        
        switch mopubError {
        case MOPUBErrorNoNetworkData:
            fallthrough
        case MOPUBErrorNoInventory:
            fallthrough
        case MOPUBErrorAdapterHasNoInventory:
            // No data found in NSHTTPURL response
            eventErr = eventError(eventErr, code: POBErrorCode.errorNoAds)
            
        case MOPUBErrorNetworkTimedOut:
            eventErr = eventError(eventErr, code: POBErrorCode.errorNetworkError)
            
        case MOPUBErrorServerError:
            eventErr = eventError(eventErr, code: POBErrorCode.errorServerError)
            
        case MOPUBErrorAdRequestTimedOut:
            eventErr = eventError(eventErr, code: POBErrorCode.errorTimeout)
            
        case MOPUBErrorAdapterInvalid:
            fallthrough
        case MOPUBErrorAdapterNotFound:
            eventErr = eventError(eventErr, code: POBErrorCode.signalingError)
            
        case MOPUBErrorAdUnitWarmingUp:
            fallthrough
        case MOPUBErrorSDKNotInitialized:
            eventErr = eventError(eventErr, code: POBErrorCode.errorInternalError)
            
        case MOPUBErrorUnableToParseJSONAdResponse:
            fallthrough
        case MOPUBErrorUnexpectedNetworkResponse:
            eventErr = eventError(eventErr, code: POBErrorCode.errorInvalidResponse)
            
        case MOPUBErrorNoRenderer:
            eventErr = eventError(eventErr, code: POBErrorCode.errorRenderError)
        default:
            break
        }
        _delegate?.failedWithError(eventErr)
    }
    
    func willPresentModalView(forAd view: MPAdView!) {
        _delegate?.willPresentModal()
    }
    
    func didDismissModalView(forAd view: MPAdView!) {
        _delegate?.didDismissModal()
    }
    
    func willLeaveApplication(fromAd view: MPAdView!) {
        _delegate?.willLeaveApp()
    }
    
    // MARK: - Helper methods
    func eventError(_ error : NSError, code : POBErrorCode) -> NSError {
        let eventError = NSError(domain: kPOBErrorDomain, code:code.rawValue, userInfo: error.userInfo)
        return eventError
    }
    
    func keywords(for bid: POBBid?) -> String? {
        var keywords = ""
        let targetingInfo = bid?.targetingInfo()
        
        for key in (targetingInfo?.keys)! {
            keywords += String(format: "%@:%@,", key as! String,targetingInfo?[key] as! String)
        }
        if ((bannerView?.keywords?.count) != nil) {
            keywords += bannerView?.keywords ?? ""
        }else{
            if keywords.count != 0 {
                if let subRange = Range<String.Index>(NSRange(location: keywords.count - 1, length: 1), in: keywords) { keywords.removeSubrange(subRange) }
            }
        }
        return keywords
    }

}
